File: 16/python3/main.py
#!/usr/bin/env python3

class Solution:
    def threeSumClosest(self, nums, target):
        closest = float('inf')  # Initialize closest sum to a very large number
        min_diff = float('inf')  # Track the minimum difference between sum and target

        # Sort the array to apply two-pointer approach
        nums.sort()

        for i in range(len(nums)):
            # Skip duplicates for the first number
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1

            while left < right:
                sum_ = nums[i] + nums[left] + nums[right]
                diff = abs(target - sum_)

                # Update the closest sum if current sum is closer to the target
                if diff < min_diff:
                    closest = sum_
                    min_diff = diff

                # Move the pointers based on how the sum compares to the target
                if sum_ < target:
                    left += 1
                elif sum_ > target:
                    right -= 1
                else:
                    # If sum matches the target exactly, return it
                    return sum_

        return closest

File: 16/cpp/main.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
#include <vector>

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int closest = INT_MAX / 100;
        int min_diff = closest;
        
        // Sort the array to allow two-pointer approach
        std::sort(nums.begin(), nums.end());
        
        for (int i = 0; i < nums.size(); i++) {
            // Skip the same element to avoid duplicate triplets
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            
            int left = i + 1;
            int right = nums.size() - 1;
            
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                int diff = std::abs(target - sum);
                
                if (diff < min_diff) {
                    // Set the closest sum
                    closest = sum;
                    min_diff = diff;
                }
                    
                // Move the pointers based on comparison with target
                if (sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                } else {
                    // If the sum is exactly the target, return it immediately
                    return sum;
                }
            }
        }
        
        return closest;
    }
};

int main() {
    Solution sol;

    // Test case X: Y
    RETURN_TYPE resultX = sol.FUNCTION_NAME(Y);
    std::cout << "Test X: " << resultX << std::endl;

    return 0;
}
